### PR TITLE
Improve connectors

### PR DESCRIPTION
**Pull request by trabucayre**
This PR try to provides a solution to comments in [PR](https://github.com/litex-hub/litex-boards/pull/418).
- the first commit add a serie of method, similar to `add_extension`, to allows extending connectors dictionary dinamicaly
- second commit add, when a connector reference another connector, a recursive way to retrieve pin name.

### CODE REVIEW

File: litex/build/generic_platform.py [hunk 1]
```diff
@@ -132,6 +132,9 @@ def _resource_type(resource):
 class ConnectorManager:
     def __init__(self, connectors):
         self.connector_table = dict()
+        self.add_connectors(connectors)
+
+    def add_connectors(self, connectors):
```
Comment (enjoy-digital) [inline litex/build/generic_platform.py:137]: Maybe we should have a `add_connector` and move the `for connector in connectors` loop in __init__?

File: litex/build/generic_platform.py [hunk 2]
```diff
@@ -193,6 +199,9 @@ def __init__(self, io, connectors):
     def add_extension(self, io):
         self.available.extend(io)
 
+    def add_connectors(self, connectors):
```
Comment (enjoy-digital) [inline litex/build/generic_platform.py:202]: Same: `add_connector`?

File: litex/build/generic_platform.py [hunk 3]
```diff
@@ -356,6 +365,9 @@ def add_platform_command(self, *args, **kwargs):
     def add_extension(self, *args, **kwargs):
         return self.constraint_manager.add_extension(*args, **kwargs)
 
+    def add_connectors(self, *args, **kwargs):
```
Comment (enjoy-digital) [inline litex/build/generic_platform.py:368]: Same: add_connector?

**Review by enjoy-digital — CHANGES_REQUESTED** [review]
Thanks @trabucayre, this will be very useful. The exposed method should however probably only add one connector at once, so call it`add_connector` and do the appropriate change (mostly move for loop).

File: litex/build/generic_platform.py (hunk 1 above)
Comment (trabucayre) [reply litex/build/generic_platform.py:137]: I think testing `connector` type is more easy, avoid the overhead to call this method multiple time and allows to pass a serie of connectors: this method may be used with a scalar (tuple) or with a list (as done for **lite** variant for the **sipeed Tang Primer 20k** dock):

```
if isinstance(connector, tuple):
    connector = [connector]
```

Thanks to that this method will be more generic and avoid `[]` boilerplate when only one connector must be added.

**Comment by trabucayre** [comment]
> Thanks @trabucayre, this will be very useful. The exposed method should however probably only add one connector at once, so call it`add_connector` and do the appropriate change (mostly move for loop).

In fact I have used `connectors` with **s** because it's called like this everywhere. But if you prefer without I will change that.
The first motivation to this method is to answer case like tang primer where 4 pmods and 2 connectors must be added.

**Comment by enjoy-digital** [comment]
Thanks @trabucayre, we could keep `connectors` in `__init__` (for retro compatibility), but I would have a preference to have a `add_connector` method.

**Comment by trabucayre** [comment]
Done

**Pull request merged**